storagemarket/impl/shardregistration.go

```diff
@@ -0,0 +1,80 @@
+package storageimpl
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-fil-markets/dagstore"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/impl/providerstates"
+)
+
+var shardRegKey = datastore.NewKey("shards-registered")
+
+type ShardRegistration struct {
+	ds       datastore.Datastore
+	dagStore dagstore.DagStoreWrapper
+}
+
+func NewShardRegistration(ds datastore.Datastore, dagStore dagstore.DagStoreWrapper) *ShardRegistration {
+	return &ShardRegistration{
+		ds:       ds,
+		dagStore: dagStore,
+	}
+}
+
+func (r *ShardRegistration) registerShards(ctx context.Context, deals []storagemarket.MinerDeal) error {
+	// Check if all shards have already been registered
+	_, err := r.ds.Get(shardRegKey)
+	if err == nil {
+		// All shards have been registered, bail out
+		return nil
+	}
+	// Expect ErrNotFound if shards have not yet been registered
+	if !xerrors.Is(err, datastore.ErrNotFound) {
+		// There was some other error (not ErrNotFound)
+		return xerrors.Errorf("failed to get shard registration status: %w", err)
+	}
+
+	// Filter for deals that are currently sealing
+	batch := r.dagStore.RegistrationBatch()
+	for _, deal := range deals {
+		if deal.Ref.PieceCid == nil {
+			continue
+		}
+
+		// Check if the deal has been handed off to the sealing subsystem
+		var sealing bool
+		for _, state := range providerstates.ProviderSealingStates {
+			if deal.State == state {
+				sealing = true
+				break
+			}
+		}
+		if !sealing {
+			continue
+		}
```

Comment:
What's the rationale for this? Will this support lazy initialization for sealed deals going forward?

Reply:
We only need to call RegisterShard for deals that have already passed the point at which the deal is handed off to the sealing subsystem.
If the deal has not yet reached that point it will be called from within the state machine once the deal reaches that point.

Reply:
Added a comment explaining this in the code